cupy/linalg/decomposition.py

```diff
@@ -118,21 +118,29 @@ def qr(a, mode='reduced'):
     mn = min(m, n)
     handle = device.get_cusolver_handle()
     dev_info = cupy.empty(1, dtype=numpy.int32)
-    # compute working space of geqrf and ormqr, and solve R
+    # compute working space of geqrf and orgqr, and solve R
     if dtype == 'f':
-        buffersize = cusolver.sgeqrf_bufferSize(handle, m, n, x.data.ptr, n)
-        workspace = cupy.empty(buffersize, dtype=numpy.float32)
-        tau = cupy.empty(mn, dtype=numpy.float32)
-        cusolver.sgeqrf(
-            handle, m, n, x.data.ptr, m,
-            tau.data.ptr, workspace.data.ptr, buffersize, dev_info.data.ptr)
-    else:  # dtype == 'd'
-        buffersize = cusolver.dgeqrf_bufferSize(handle, n, m, x.data.ptr, n)
-        workspace = cupy.empty(buffersize, dtype=numpy.float64)
-        tau = cupy.empty(mn, dtype=numpy.float64)
-        cusolver.dgeqrf(
-            handle, m, n, x.data.ptr, m,
-            tau.data.ptr, workspace.data.ptr, buffersize, dev_info.data.ptr)
+        geqrf_bufferSize = cusolver.sgeqrf_bufferSize
+        geqrf = cusolver.sgeqrf
+    elif dtype == 'd':
+        geqrf_bufferSize = cusolver.dgeqrf_bufferSize
+        geqrf = cusolver.dgeqrf
+    elif dtype == 'F':
+        geqrf_bufferSize = cusolver.cgeqrf_bufferSize
+        geqrf = cusolver.cgeqrf
+    elif dtype == 'D':
+        geqrf_bufferSize = cusolver.zgeqrf_bufferSize
+        geqrf = cusolver.zgeqrf
+    else:
+        msg = ('dtype must be float32, float64, complex64 or complex128'
+               ' (actual: {})'.format(a.dtype))
+        raise ValueError(msg)
```

Comment:
Nitpick: The block is unreachable.  (See https://github.com/cupy/cupy/pull/2509#discussion_r331863033)